tools/settings/trigger_settings.txt
0,102,0
0,103,6
0,104,6
0,105,7
0,106,14
0,107,24
0,118,6
0,109,0
0,111,3
0,119,15
0,114,0
0,112,38
0,113,40
0,116,127
0,117,127
0,121,20
0,122,10
0,120,0
0,110,0
1,102,6
1,103,20
1,104,6
1,105,9
1,106,14
1,107,12
1,118,12
1,109,4
1,111,0
1,119,15
1,114,0
1,112,36
1,113,36
1,116,127
1,117,127
1,121,9
1,122,14
1,120,0
1,110,0
2,102,13
2,103,8
2,104,4
2,105,9
2,106,14
2,107,12
2,118,6
2,109,0
2,111,0
2,119,15
2,114,0
2,112,22
2,113,22
2,116,26
2,117,26
2,121,25
2,122,15
2,120,0
2,110,0
3,102,3
3,103,9
3,104,11
3,105,5
3,106,0
3,107,12
3,118,6
3,109,0
3,111,0
3,119,15
3,114,0
3,112,44
3,113,44
3,116,127
3,117,127
3,121,25
3,122,15
3,120,0
3,110,0
4,102,9
4,103,10
4,104,8
4,105,9
4,106,14
4,107,10
4,118,6
4,109,4
4,111,1
4,119,0
4,114,7
4,112,49
4,113,55
4,116,127
4,117,127
4,121,9
4,122,14
4,120,0
4,110,0
5,102,13
5,103,8
5,104,4
5,105,9
5,106,14
5,107,12
5,118,6
5,109,0
5,111,0
5,119,15
5,114,0
5,112,48
5,113,50
5,116,127
5,117,127
5,121,25
5,122,15
5,120,0
5,110,0
6,102,2
6,103,10
6,104,7
6,105,26
6,106,11
6,107,22
6,118,7
6,109,0
6,111,3
6,119,15
6,114,0
6,112,51
6,113,53
6,116,127
6,117,127
6,121,25
6,122,15
6,120,0
6,110,0
7,102,13
7,103,8
7,104,4
7,105,9
7,106,14
7,107,12
7,118,6
7,109,0
7,111,0
7,119,15
7,114,0
7,112,45
7,113,47
7,116,127
7,117,127
7,121,25
7,122,15
7,120,0
7,110,0
8,102,1
8,103,8
8,104,1
8,105,9
8,106,15
8,107,10
8,118,6
8,109,0
8,111,3
8,119,15
8,114,0
8,112,43
8,113,58
8,116,127
8,117,127
8,121,14
8,122,15
8,120,0
8,110,0
9,102,8
9,103,22
9,104,15
9,105,9
9,106,14
9,107,12
9,118,61
9,109,0
9,111,1
9,119,0
9,114,0
9,112,59
9,113,52
9,116,127
9,117,127
9,121,25
9,122,15
9,120,0
9,110,0


ChangeLog.md
# Edrumulus Project Log


## 2022-07-23 Confusion about the PDA-120LS pad

  My assumption was that all Roland pads with mesh heads do support positional sensing. So, I
  acciently recommended the PDA-120LS for having positional sensing support with Edrumulus.
  After recording a test signal with that pad, I had trouble to find any parameters where the
  positional sensing works at all. Finally, I found out that this pad does not use a center-
  mounted pizeo but three piezos at the edges of the mesh head. That explains why I could not
  get the positional sensing working. So, be careful what kind of mesh head pad you buy if you
  want to have positional sensing.

## 2022-07-02 Youtube video of prototype 3 (HD-1)

  A [Youtube Video](https://youtu.be/tBlKvivwAE4) of Edrumulus prototype 3 (HD-1) is
  now available showing the trigger performance in real action.

## 2022-06-18 Unreliable settings storage on ESP32

  Unfortunately, it seems that the settings storage on the ESP32 does not work as expected.
  On my prototype 3 (HD-1), some settings are not recalled correctly. This especially
  happens if the attached Raspberry Pi must be resetted after, e.g., a freeze. Either the
  sudden power off of the ESP32 is the cause or the Raspberry Pi sends arbitrary MIDI signals
  when it is shut-down. This has to be investigated...

## 2022-01-22 [[r0.4](https://github.com/corrados/edrumulus/releases/tag/r0_4)] Store settings on Edrumulus

  Edrumulus stores all trigger parameter on the micro processor using the Arduino
  EEPROM library now. It is no longer necessary to hard-code pad trigger parameters for
  your prototype in edrumulus.ino. After flashing the new firmware for the first time,
  the settings must be reset. Either use the Linux EdrumulusGUI and press the **r** key
  or use the Octave edrumuluscontrol.m GUI and press the **Reset All Settings** button.

## 2022-01-16 [[r0.3](https://github.com/corrados/edrumulus/releases/tag/r0_3)] New Linux console GUI

  There is a new tool available to show/modify Edrumulus parameters in a Linux (remote) shell
  using the ncurses library. This comes handy if Edrumulus is used with Drumgizmo running on
  a headless Raspberry Pi. This GUI is coded in one single file tools/edrumulus_gui.cpp.
  <br/>![Edrumulus console GUI](algorithm/images/edrumulus_console_gui.jpg)

## 2022-01-09 Edrumulus prototype 3 (HD-1)

  Today was the first time I tested Edrumulus in a full drum-kit setup and
  stand-alone (i.e. no PC or Laptop needed). The prototype 3 consists of an
  ESP32 with 7 pad and 1 control inputs, a Raspberry Pi 3 with a 32 GB SD card
  and a Behringer UFO 202 USB audio interface. It worked pretty amazingly well.
  Of course, there is plenty of room for improvement. But for a first try it was
  a great experience. Even positional sensing was possible with the PDX-8
  mesh pad even that it does not have a center mounted piezo trigger.
  <br/>![Prototype 3 (HD-1)](algorithm/images/edrumulus_prototype3_hd1.jpg)

## 2022-01-09 Possible out-of-memory issue for the ESP32

  Today I flashed the current Edrumulus code on my ESP32 and it immediately crashed. After
  disabling the initialization of the last configured pad it did not crash anymore so, I
  assume that we have an out-of-memory issue for the ESP32.

## 2022-01-08 Hot spot detection for mesh head pads

  I was working on a hot spot detection algorithm on a Git side branch recently. As a reference,
  I have used the piezo signal of a PD-80R and PD120 recorded with a conventional audio card.
  These audio card inputs are usually optimized for low impedence microphones. I developed an
  algorithm which worked fine for these signals but did not work at all for running it in
  real-time on a micro controller. The reason for that is caused by the mismatched input of the
  audio card for piezo signals. The shape of the signal is so different that the algorithm
  did not work at all. A new algorithm and further investigations are needed.

## 2021-12-16 Rim shot detection again

  After fixing a bug with the detection of the first peak and introducing a band-pass filter
  for the rim piezo signal of the PD-80R, the rim shot detection is now acceptable (but still
  far from perfect).

## 2021-12-05 New algorithms ready

  The improved peak detection and positional sensing algorithms are now ready and the code
  was merged from a Git side branch to the main branch. By implementing these improvements,
  the source code was heavily changed. Thus, a lot of fine tuning has to be done to, e.g.,
  tweak the pad parameters for the new algorithms and fix possible new issues.

## 2021-11-21 Time to say goodbye, Hilbert...

  The filtered signal currently used for peak detection is a Hilbert filtered signal with
  a moving average low-pass filter applied to it. The initial motivation for this design
  was to reduce the noise and improve the main lobe energy measurement. It turned out that
  it is advantageous to just use the peak value instead of an average.<br/>
  Therefore, we should use an optimized filter which improves the peak detection. To
  reduce the noise floor, a low-pass filter should still be applied. At the same time, some
  high pass filter should be applied to filter out very low frequency noise caused by, e.g,
  the movement of the pad stand after a hit on the pad. Thus, a Butterworth band-pass filter
  with a pass-band of 40 to 400 Hz was chosen. The shorter the pass-band, the longer the filter
  settle time will be. Since the settle time should be as short as possible, it
  is a trade-off between noise reduction and latency.<br/>
  In the following picture, the current Hilbert filter based design is compared to the new
  Butterworth filter design where the peak detection threshold is tweaked for each scenario
  so that no false peaks are detected in the entire signal:

parameters.cpp
// Copyright (c) 2020-2026 Volker Fischer
// SPDX-License-Identifier: GPL-2.0-or-later

#include "edrumulus.h"

void Pad::apply_preset_pad_settings()
{
  // apply PRESET settings (might be overwritten by pad-specific properties)
  pad_settings.velocity_threshold        = 8;  // 0..31
  pad_settings.velocity_sensitivity      = 9;  // 0..31
  pad_settings.mask_time_ms              = 6;  // 0..31 (ms)
  pad_settings.pos_threshold             = 9;  // 0..31
  pad_settings.pos_sensitivity           = 14; // 0..31
  pad_settings.rim_pos_threshold         = 25; // 0..31
  pad_settings.rim_pos_sensitivity       = 15; // 0..31
  pad_settings.rim_shot_threshold        = 12; // 0..31
  pad_settings.rim_shot_boost            = 15; // 0..31
  pad_settings.cancellation              = 0;  // 0..31
  pad_settings.coupled_pad_idx           = 0;  // 0..[number of pads - 1] (0 means disabled)
  pad_settings.curve_type                = LINEAR;
  pad_settings.is_control                = false;  // default is a normal pad
  pad_settings.is_rim_switch             = false;  // default is a piezo for the rim sensor
  pad_settings.pos_sense_is_used         = false;  // must be explicitely enabled if it shall be used
  pad_settings.rim_shot_is_used          = false;  // must be explicitely enabled if it shall be used
  pad_settings.first_peak_diff_thresh_db = 8.0f;   // pad specific parameter: allowed difference between first peak and later peak in scan time
  pad_settings.mask_time_decay_fact_db   = 15.0f;  // pad specific parameter: during mask time decay factor for maximum peak in scan time
  pad_settings.scan_time_ms              = 2.5f;   // pad specific parameter: scan time after first detected peak
  pad_settings.pre_scan_time_ms          = 2.5f;   // pad specific parameter: pre-scan time to search for first peak
  pad_settings.decay_est_delay_ms        = 7.0f;   // pad specific parameter: delay after second main peak until decay power estimation starts
  pad_settings.decay_est_len_ms          = 4.0f;   // pad specific parameter: decay power estimation window length
  pad_settings.decay_est_fact_db         = 16.0f;  // pad specific parameter: decay power estimation factor (to get over decay ripple)
  pad_settings.decay_fact_db             = 1.0f;   // pad specific parameter: vertical shift of the decay function in dB
  pad_settings.decay_len1_ms             = 0.0f;   // pad specific parameter: length of the decay 1
  pad_settings.decay_grad_fact1          = 200.0f; // pad specific parameter: decay function gradient factor 1
  pad_settings.decay_len2_ms             = 350.0f; // pad specific parameter: length of the decay 2
  pad_settings.decay_grad_fact2          = 200.0f; // pad specific parameter: decay function gradient factor 2
  pad_settings.decay_len3_ms             = 0.0f;   // pad specific parameter: length of the decay 3
  pad_settings.decay_grad_fact3          = 200.0f; // pad specific parameter: decay function gradient factor 3
  pad_settings.pos_low_pass_cutoff       = 150.0f; // pad specific parameter: low-pass filter cut-off in Hz for positional sensing
  pad_settings.pos_invert                = false;  // pad specific parameter: invert the positional sensing metric
  pad_settings.rim_use_low_freq_bp       = true;   // pad specific parameter: use low frequency band-pass filter for rim shot detection
  pad_settings.rim_shot_window_len_ms    = 3.5f;   // pad specific parameter: window length for rim shot detection
  pad_settings.clip_comp_ampmap_step     = 0.08f;  // pad specific parameter: clipping compensation amplitude mapping step, conservative value from PD80R as default

  switch (pad_settings.pad_type)
  {
    // Mesh pads ---------------------------------------------------------------
    case PD120: // dual trigger
      pad_settings.velocity_threshold   = 6;
      pad_settings.velocity_sensitivity = 6;
      pad_settings.pos_threshold        = 7;
      pad_settings.pos_sensitivity      = 14;
      pad_settings.rim_pos_threshold    = 20;
      pad_settings.rim_pos_sensitivity  = 10;
      pad_settings.rim_shot_threshold   = 24;
      pad_settings.pos_sense_is_used    = true;
      pad_settings.rim_shot_is_used     = true;
      break;

    case PD80R: // dual trigger
      pad_settings.velocity_sensitivity = 1;
      pad_settings.rim_shot_threshold   = 10;
      pad_settings.pos_threshold        = 9;
      pad_settings.pos_sensitivity      = 15;
      pad_settings.rim_pos_threshold    = 14;
      pad_settings.rim_pos_sensitivity  = 15;
      pad_settings.scan_time_ms         = 3.0f;
      pad_settings.decay_len2_ms        = 75.0f;
      pad_settings.decay_grad_fact2     = 300.0f;
      pad_settings.decay_len3_ms        = 300.0f;
      pad_settings.decay_grad_fact3     = 100.0f;
      pad_settings.rim_use_low_freq_bp  = false;
      pad_settings.pos_sense_is_used    = true;
      pad_settings.rim_shot_is_used     = true;
      break;

    case PDA120LS: // dual trigger
      pad_settings.velocity_threshold        = 11;
      pad_settings.velocity_sensitivity      = 7;
      pad_settings.pos_threshold             = 31;
      pad_settings.pos_sensitivity           = 16;
      pad_settings.rim_shot_threshold        = 30;
      pad_settings.decay_grad_fact2          = 250.0f;
      pad_settings.decay_fact_db             = 5.0f;
      pad_settings.pre_scan_time_ms          = 3.5f;
      pad_settings.first_peak_diff_thresh_db = 7.0f;
      pad_settings.pos_invert                = true;
      pad_settings.pos_sense_is_used         = true;
      pad_settings.rim_shot_is_used          = true;
      break;

    case PDX100:
      pad_settings.velocity_threshold   = 6;
      pad_settings.velocity_sensitivity = 5;
      pad_settings.pos_threshold        = 9;
      pad_settings.pos_sensitivity      = 16;
      pad_settings.rim_shot_threshold   = 24;
      pad_settings.decay_fact_db        = 2.0f;
      pad_settings.decay_len2_ms        = 65.0f;
      pad_settings.decay_grad_fact2     = 300.0f;
      pad_settings.decay_len3_ms        = 350.0f;
      pad_settings.decay_grad_fact3     = 150.0f;
      pad_settings.pos_low_pass_cutoff  = 120.0f;
      pad_settings.pos_sense_is_used    = true;
      pad_settings.rim_shot_is_used     = true;
      break;

    case PDX8:
      pad_settings.velocity_threshold   = 6;
      pad_settings.velocity_sensitivity = 4;
      pad_settings.rim_shot_threshold   = 14;
      pad_settings.pos_threshold        = 21;
      pad_settings.pos_sensitivity      = 27;
      pad_settings.pos_sense_is_used    = true;
      pad_settings.rim_shot_is_used     = true;
      break;

    case DIABOLO12:
      pad_settings.scan_time_ms              = 2.0f;
      pad_settings.mask_time_ms              = 8.0f;
      pad_settings.first_peak_diff_thresh_db = 13.0f;
      pad_settings.mask_time_decay_fact_db   = 20.0f;
      pad_settings.decay_grad_fact2          = 270.0f;
      pad_settings.decay_fact_db             = 6.0f;
      pad_settings.decay_est_delay_ms        = 20.0f;
      pad_settings.pos_low_pass_cutoff       = 50.0f; // positional sensing seems not to work correctly
      pad_settings.pos_sense_is_used         = true;
      pad_settings.rim_shot_is_used          = true;
      break;

    case MPS750X_SNARE: // dual trigger
      pad_settings.scan_time_ms         = 6.0f;
      pad_settings.pre_scan_time_ms     = 3.5f;
      pad_settings.decay_grad_fact2     = 150.0f;
      pad_settings.decay_len2_ms        = 450.0f;
      pad_settings.rim_use_low_freq_bp  = false;
      pad_settings.rim_shot_is_used     = true;
      pad_settings.rim_shot_threshold   = 12;
      pad_settings.velocity_threshold   = 2;
      pad_settings.velocity_sensitivity = 6;
      pad_settings.pos_sense_is_used    = true;
      pad_settings.pos_sensitivity      = 3;
      break;

    case MPS750X_TOM: // dual trigger
      pad_settings.scan_time_ms         = 6.0f;
      pad_settings.pre_scan_time_ms     = 3.5f;
      pad_settings.decay_grad_fact2     = 150.0f;
      pad_settings.decay_len2_ms        = 450.0f;
      pad_settings.rim_shot_is_used     = false;
      pad_settings.velocity_threshold   = 2;
      pad_settings.velocity_sensitivity = 3;
      break;

    // Rubber pads -------------------------------------------------------------
    case PD5: // single trigger
      pad_settings.scan_time_ms     = 4.0f;
      pad_settings.mask_time_ms     = 9.0f;
      pad_settings.decay_grad_fact2 = 400.0f;
      break;

    case PD6: // single trigger
      pad_settings.velocity_sensitivity = 4;
      pad_settings.scan_time_ms         = 1.5f;
      pad_settings.decay_grad_fact2     = 400.0f;
      break;

    case PD8: // dual trigger
      pad_settings.is_rim_switch         = true;
      pad_settings.velocity_sensitivity  = 3;
      pad_settings.pos_threshold         = 26;
      pad_settings.pos_sensitivity       = 11;
      pad_settings.rim_shot_threshold    = 22;
      pad_settings.mask_time_ms          = 7;
      pad_settings.scan_time_ms          = 1.3f;
      pad_settings.decay_est_delay_ms    = 6.0f;
      pad_settings.decay_fact_db         = 5.0f;
      pad_settings.decay_len2_ms         = 30.0f;
      pad_settings.decay_grad_fact2      = 600.0f;
      pad_settings.decay_len3_ms         = 150.0f;
      pad_settings.decay_grad_fact3      = 120.0f;
      pad_settings.pos_sense_is_used     = true;
      pad_settings.rim_shot_is_used      = true;
      pad_settings.clip_comp_ampmap_step = 0.4f;
      break;

    case TP80: // single trigger
      pad_settings.velocity_sensitivity = 8;
      pad_settings.pos_threshold        = 27;
      pad_settings.pos_sensitivity      = 19;
      pad_settings.scan_time_ms         = 2.75f;
      pad_settings.decay_est_delay_ms   = 11.0f;
      pad_settings.decay_len2_ms        = 60.0f;
      pad_settings.decay_grad_fact2     = 400.0f;
      pad_settings.decay_len3_ms        = 700.0f;
      pad_settings.decay_grad_fact3     = 60.0f;
      pad_settings.pos_invert           = true;
      pad_settings.pos_sense_is_used    = true;
      break;

    case HD1TOM: // single trigger
      pad_settings.velocity_threshold   = 8;
      pad_settings.velocity_sensitivity = 5;
      pad_settings.curve_type           = LOG1;
      pad_settings.scan_time_ms         = 1.5f;
      pad_settings.decay_grad_fact2     = 300.0f;
      pad_settings.pos_low_pass_cutoff  = 300.0f;
      break;

    // Kick drum pads ----------------------------------------------------------
    case KD7: // single trigger
      pad_settings.velocity_threshold   = 11;
      pad_settings.velocity_sensitivity = 6;
      pad_settings.curve_type           = LOG2; // less dynamic on kick (similar to other drum modules)
      pad_settings.scan_time_ms         = 2.0f;
      pad_settings.decay_est_delay_ms   = 8.0f;
      pad_settings.decay_fact_db        = 5.0f;
      pad_settings.decay_len1_ms        = 4.0f;
      pad_settings.decay_grad_fact1     = 30.0f;
      pad_settings.decay_len2_ms        = 30.0f;
      pad_settings.decay_grad_fact2     = 450.0f;
      pad_settings.decay_len3_ms        = 500.0f;
      pad_settings.decay_grad_fact3     = 45.0f;
      break;

    case KD8: // single trigger
      pad_settings.velocity_sensitivity    = 2;
      pad_settings.curve_type              = LOG2; // less dynamic on kick (similar to other drum modules)
      pad_settings.scan_time_ms            = 3.0f;
      pad_settings.mask_time_decay_fact_db = 10.0f;
      pad_settings.decay_grad_fact2        = 450.0f;
      pad_settings.decay_len3_ms           = 500.0f;
      pad_settings.decay_grad_fact3        = 45.0f;
      break;

    case KT10: // single trigger
      pad_settings.velocity_sensitivity = 6;
      pad_settings.scan_time_ms         = 2.0f;
      pad_settings.decay_est_delay_ms   = 8.0f;
      pad_settings.decay_fact_db        = 5.0f;
      pad_settings.decay_len1_ms        = 4.0f;
      pad_settings.decay_grad_fact1     = 30.0f;
      pad_settings.decay_len2_ms        = 30.0f;
      pad_settings.decay_grad_fact2     = 350.0f;
      pad_settings.decay_len3_ms        = 500.0f;
      pad_settings.decay_grad_fact3     = 100.0f;
      break;

    case KD120: // single trigger
      pad_settings.scan_time_ms      = 6.0f;
      pad_settings.decay_fact_db     = 12.0f;
      pad_settings.decay_est_fact_db = 20.0f;
      break;

    case MPS750X_KICK: // single trigger
      pad_settings.first_peak_diff_thresh_db = 3.0f;
      pad_settings.scan_time_ms              = 6.0f;
      pad_settings.pre_scan_time_ms          = 3.5f;
      pad_settings.decay_grad_fact2          = 150.0f;
      pad_settings.decay_len2_ms             = 450.0f;
      pad_settings.velocity_threshold        = 10;
      pad_settings.velocity_sensitivity      = 10;
      pad_settings.curve_type                = LOG2;
      break;

    // Cymbal pads -------------------------------------------------------------
    case HD120: // dual trigger (advertised as single trigger)
      pad_settings.is_rim_switch        = true;
      pad_settings.scan_time_ms         = 4.5f;
      pad_settings.decay_grad_fact2     = 90.0f;
      pad_settings.decay_len2_ms        = 500.0f;
      pad_settings.rim_use_low_freq_bp  = false;
      pad_settings.rim_shot_is_used     = true;
      pad_settings.rim_shot_threshold   = 24;
      pad_settings.rim_shot_boost       = 0;
      pad_settings.velocity_threshold   = 8;
      pad_settings.velocity_sensitivity = 4;
      pad_settings.curve_type           = EXP2;
      break;

    case CY5: // dual trigger
      pad_settings.is_rim_switch        = true;
      pad_settings.velocity_threshold   = 6;
      pad_settings.velocity_sensitivity = 4;
      pad_settings.rim_shot_threshold   = 12;
      pad_settings.rim_shot_boost       = 0;
      pad_settings.scan_time_ms         = 3.0f;
      pad_settings.mask_time_ms         = 8.0f;
      pad_settings.decay_fact_db        = 3.0f;
      pad_settings.rim_shot_is_used     = true;
      break;

    case CY6: // dual trigger
      pad_settings.is_rim_switch        = true;
      pad_settings.velocity_sensitivity = 6;
      pad_settings.rim_shot_threshold   = 12;
      pad_settings.rim_shot_boost       = 0;
      pad_settings.scan_time_ms         = 6.0f;
      pad_settings.decay_len2_ms        = 150.0f;
      pad_settings.decay_grad_fact2     = 120.0f;
      pad_settings.decay_len3_ms        = 450.0f;
      pad_settings.decay_grad_fact3     = 30.0f;
      pad_settings.rim_shot_is_used     = true;
      break;

    case CY8: // dual trigger
      pad_settings.is_rim_switch        = true;
      pad_settings.velocity_threshold   = 10;
      pad_settings.velocity_sensitivity = 5;
      pad_settings.rim_shot_threshold   = 10;
      pad_settings.rim_shot_boost       = 0;
      pad_settings.curve_type           = LOG2;
      pad_settings.scan_time_ms         = 6.0f;
      pad_settings.decay_len1_ms        = 10.0f;
      pad_settings.decay_grad_fact1     = 10.0f;
      pad_settings.decay_len2_ms        = 100.0f;
      pad_settings.decay_grad_fact2     = 200.0f;
      pad_settings.decay_len3_ms        = 450.0f;
      pad_settings.decay_grad_fact3     = 30.0f;
      pad_settings.rim_shot_is_used     = true;
      break;

    case VH12: // dual trigger
               // TODO if the Hi-Hat is open just a little bit, we get double triggers
      pad_settings.is_rim_switch        = true;
      pad_settings.velocity_sensitivity = 5;
      pad_settings.rim_shot_threshold   = 23;
      pad_settings.rim_shot_boost       = 0;
      pad_settings.scan_time_ms         = 4.0f;
      pad_settings.decay_est_delay_ms   = 9.0f;
      pad_settings.decay_fact_db        = 5.0f;
      pad_settings.decay_len2_ms        = 27.0f;
      pad_settings.decay_grad_fact2     = 700.0f;
      pad_settings.decay_len3_ms        = 600.0f; // must be long because of open Hi-Hat ringing
      pad_settings.decay_grad_fact3     = 75.0f;
      break;

    case LEHHS12C: // dual trigger (Lemon HHC12)
      pad_settings.is_rim_switch        = true;
      pad_settings.scan_time_ms         = 4.0f;
      pad_settings.decay_fact_db        = 5.0f;
      pad_settings.decay_len2_ms        = 600.0f;
      pad_settings.decay_grad_fact2     = 100.0f;
      pad_settings.rim_shot_is_used     = true;
      pad_settings.rim_shot_threshold   = 25;
      pad_settings.rim_shot_boost       = 0;
      pad_settings.velocity_threshold   = 18;
      pad_settings.velocity_sensitivity = 6;
      break;

    case MPS750X_RIDE: // 3-zone
      pad_settings.is_rim_switch        = true;
      pad_settings.scan_time_ms         = 3.0f;
      pad_settings.decay_est_delay_ms   = 1.0f;
      pad_settings.decay_grad_fact2     = 150.0f;
      pad_settings.decay_len2_ms        = 450.0f;
      pad_settings.rim_use_low_freq_bp  = false;
      pad_settings.rim_shot_is_used     = true;
      pad_settings.rim_shot_threshold   = 18;
      pad_settings.rim_shot_boost       = 0;
      pad_settings.velocity_threshold   = 12;
      pad_settings.velocity_sensitivity = 6;
      break;

    case MPS750X_CRASH: // 2-zone
      pad_settings.is_rim_switch        = true;
      pad_settings.scan_time_ms         = 3.0f;
      pad_settings.decay_est_delay_ms   = 1.0f;
      pad_settings.decay_grad_fact2     = 150.0f;
      pad_settings.decay_len2_ms        = 450.0f;
      pad_settings.rim_use_low_freq_bp  = false;
      pad_settings.rim_shot_is_used     = true;
      pad_settings.rim_shot_threshold   = 20;
      pad_settings.rim_shot_boost       = 0;
      pad_settings.velocity_threshold   = 12;
      pad_settings.velocity_sensitivity = 6;
      break;

    // Hi-hat controllers ------------------------------------------------------
    case FD8:
      pad_settings.is_control           = true;
      pad_settings.pos_threshold        = 5;
      pad_settings.pos_sensitivity      = 0;
      pad_settings.velocity_threshold   = 9;  // pedal stomp configuration
      pad_settings.velocity_sensitivity = 11; // pedal stomp configuration
      break;

    case VH12CTRL:
      pad_settings.is_control      = true;
      pad_settings.pos_threshold   = 19;
      pad_settings.pos_sensitivity = 28;
      break;

    case LEHHS12C_CTRL: // 0-35k poti
      pad_settings.is_control           = true;
      pad_settings.pos_threshold        = 6;
      pad_settings.pos_sensitivity      = 23;
      pad_settings.velocity_threshold   = 0;
      pad_settings.velocity_sensitivity = 22;
      break;
  }
}
